### PR TITLE
Update inline doc for types

### PR DESCRIPTION
**Pull request by ekexium**
Added some inline doc to types.
Removed a few unused or unnecessary code.

Related issue: #194

### CODE REVIEW

**Comment by ekexium** [comment]
@nrc PTAL

File: tikv-client-common/src/errors.rs [hunk 1]
```diff
@@ -189,7 +187,8 @@ impl From<tikv_client_proto::kvrpcpb::KeyError> for Error {
     }
 }
 
-/// A result holding an [`Error`](Error).
+/// The result type used in tikv-client.
+/// It holds an error type that contains varisous kinds of error that can happen in the crate.
```
Comment (nrc) [inline tikv-client-common/src/errors.rs:191]: I don't think we need either of these comments, they don't add much.

File: src/transaction/transaction.rs [hunk 2]
```diff
@@ -346,6 +400,7 @@ impl Transaction {
 /// The default TTL of a lock in milliseconds
 const DEFAULT_LOCK_TTL: u64 = 3000;
 
+/// A struct wrapping the details of 2PC.
```
Comment (nrc) [inline src/transaction/transaction.rs:403]: Could you expand this please?

File: src/transaction/transaction.rs [hunk 3]
```diff
@@ -316,7 +368,9 @@ impl Transaction {
         .await
     }
 
-    /// Pessimistically lock the keys
+    /// Pessimistically lock the keys.
+    ///
+    /// Once resovled it acquires a mutex on the key in TiKV.
```
Comment (nrc) [inline src/transaction/transaction.rs:373]: I think it is not really a mutex, just a logical lock

File: src/transaction/transaction.rs [hunk 4]
```diff
@@ -271,8 +314,16 @@ impl Transaction {
         Ok(())
     }
 
-    /// Locks the given keys.
+    /// Lock the given keys without mutating value (at the time of commit).
     ///
+    /// In optimistic mode, write conflicts are not checked until commit.
+    /// So use this command to indicate that
+    /// "I do not want to commit if the value associated with this key has been modified".
+    /// It's useful to avoid *Write Skew* anomaly.
```
Comment (nrc) [inline src/transaction/transaction.rs:322]:
```suggestion
    /// It's useful to avoid *write skew* anomaly.
```

File: src/transaction/transaction.rs [hunk 5]
```diff
@@ -271,8 +314,16 @@ impl Transaction {
...
+    ///
+    /// In pessimistic mode, please use [`get_for_update`](Transaction::get_for_update) instead.
```
Comment (nrc) [inline src/transaction/transaction.rs:324]: Do we use this command for pessimistic transactions in TiDB?

File: src/transaction/mod.rs [hunk 6]
```diff
@@ -2,12 +2,11 @@
...
 //! Transactional related functionality.
 //!
-//! Using the [`TransactionClient`](TransactionClient) you can utilize TiKV's transactional interface.
+//! Using the [`TransactionClient`](client::Client) you can utilize TiKV's transactional interface.
 //!
 //! This interface offers SQL-like transactions on top of the raw interface.
 //!
-//! **Warning:** It is not advisable to use both raw and transactional functionality in the same keyspace.
-//!
+//! **Warning:** It is not advisable to use both raw and transactional functionalities in the same keyspace.
```
Comment (nrc) [inline src/transaction/mod.rs:9]: I think the original is correct

File: src/transaction/client.rs [hunk 7]
```diff
@@ -15,7 +15,23 @@ use tikv_client_proto::{kvrpcpb, pdpb::Timestamp};
...
+/// and ends with a `rollback` or `commit` request.
+/// If a `Transaction` is dropped before it's rolled back or committed, it is automatically rolled back.
+///
+/// Transaction supports optimistic and pessimistic modes, for mroe deatils, check our
+/// [SIG-transaction](https://github.com/tikv/sig-transaction/tree/master/doc/tikv#optimistic-and-pessimistic-transactions).
+///
+/// Besides transaction, the client provides some utility methods:
+/// - `gc`: execute a GC process which clear stale data. It is not stablized yet.
+/// - `current_timestamp`: get the current `Timestamp`.
+/// - `snapshot`: get the [`Snapshot`](crate::transaction::Snapshot) of the database at a certain timestamp.
```
Comment (nrc) [inline src/transaction/client.rs:31]: Are these public?

File: src/timestamp.rs [hunk 8]
```diff
@@ -10,8 +10,13 @@ pub use tikv_client_proto::pdpb::Timestamp;
 const PHYSICAL_SHIFT_BITS: i64 = 18;
 const LOGICAL_MASK: i64 = (1 << PHYSICAL_SHIFT_BITS) - 1;
 
+/// A helper trait to convert between [`Timestamp`](Timestamp) and an u64.
+///
+/// A `Timestamp` (64 bits) contains a physical part (first 46 bits) and a logical part (last 18 bits).
 pub trait TimestampExt {
+    /// Combine physical and logical parts to get a single `Timestamp` (version).
     fn version(&self) -> u64;
+    /// Decompose the `Timestamp` (version) into physical and logical parts.
```
Comment (nrc) [inline src/timestamp.rs:19]: I think these two comments are incorrect - the functions convert to and from the u64 representation of a timestamp, to or from whatever implements this trait.

**Review by nrc — COMMENTED** [review]
Looks good! A few comments/questions inline, nothing major.

File: src/transaction/transaction.rs (hunk 3 above)
Comment (ekexium) [reply src/transaction/transaction.rs:373]: Logically it's a mutually exclusive lock 😂. But yes it might be confusing.

File: src/transaction/transaction.rs (hunk 5 above)
Comment (ekexium) [reply src/transaction/transaction.rs:324]: I think we don't. `Lock` is used for `SELECT FOR UPDATE` in optimistic transactions.

File: src/transaction/client.rs (hunk 7 above)
Comment (ekexium) [reply src/transaction/client.rs:31]: They are

File: src/transaction/transaction.rs (hunk 5 above)
Comment (nrc) [reply src/transaction/transaction.rs:324]: I thought we could only select for update in pessimistic transactions, I'm probably wrong.

File: src/transaction/client.rs (hunk 7 above)
Comment (nrc) [reply src/transaction/client.rs:31]: Do they need to be?

Comment (ekexium) [reply src/transaction/client.rs:31]: For now, I think they are intended to be public because they are not used by other parts.
`snapshot` is a wrapper of transaction to make it read-only. If it's not public we can remove it.
`gc` needs to be manually triggered by users and requires a `Timestamp`.
`current_timestamp` is the easiest way for users to get a timestamp. We don't provide convenience methods for users to calculate Timestamps. Maybe we should add these in the future.

File: src/transaction/transaction.rs (hunk 5 above)
Comment (ekexium) [reply src/transaction/transaction.rs:325]: Makes sense.

File: src/transaction/client.rs (hunk 7 above)
Comment (nrc) [reply src/transaction/client.rs:31]: Do users ever need to supply timestamps? I thought they were internal to the transaction? I guess we need to consider the lower level API.

Does Snapshot offer anything that Transaction doesn't?

File: src/transaction/transaction.rs [hunk 9]
```diff
@@ -336,6 +337,9 @@ impl Transaction {
     /// # });
     /// ```
     pub async fn lock_keys(&self, keys: impl IntoIterator<Item = impl Into<Key>>) -> Result<()> {
+        if self.is_pessimistic {
+            return Err(ErrorKind::InvalidTransactionType.into());
+        }
```
Comment (nrc) [inline src/transaction/transaction.rs:342]: This change should be in a separate PR, I think

File: src/transaction/client.rs (hunk 7 above)
Comment (ekexium) [reply src/transaction/client.rs:31]: Timestamps are needed for GC and snapshot. 

In GC users should supply a safe point. In a TiDB cluster there will be only one node that performs GC and it controls when to do it. In a TiKV client we cannot guarantee this, so I think it should be the user's choice.

`Snapshot` lets you read from any given timestamp, while `Transaction` cannot. Also the timestamp is related to GC safe point, which could be another supportive reason for users providing safe point for GC.

File: src/transaction/transaction.rs (hunk 9 above)
Comment (ekexium) [reply src/transaction/transaction.rs]: OK

File: src/transaction/client.rs (hunk 7 above)
Comment (nrc) [reply src/transaction/client.rs:31]: OK, thanks!
